### ISSUE TITLE
[receiver/filelog] Improve receiver logging and monitoring metrics

### ISSUE THREAD

**Issue by ChrsMark** [open-telemetry/opentelemetry-collector-contrib#31256]
### Component(s)

`receiver/filelog`

### Is your feature request related to a problem? Please describe.

At the moment OTel Collector can collect logs through the filelog receiver.
However the receiver does not log enough details about the file tracking actions nor extensive summary metrics.

### Describe the solution you'd like

OTel Collector should be capable to offer the following:

1) Sufficient logging when files are discovered, moved, truncated, etc.: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33237
2) Observable metrics about its current state: how many files are opened, being harvested, stale, etc.: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31544

### Additional context

Sample logs:

```console
log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]}	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:55.586+0200	debug	fileconsumer/file.go:135	matched files%!(EXTRA zapcore.Field={paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]})	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:55.586+0200	debug	fileconsumer/file.go:167	Consuming files{paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]}	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:55.786+0200	debug	fileconsumer/file.go:135	matched files%!(EXTRA zapcore.Field={paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]})	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:55.786+0200	debug	fileconsumer/file.go:167	Consuming files{paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]}	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:55.986+0200	debug	fileconsumer/file.go:135	matched files%!(EXTRA zapcore.Field={paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]})	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:55.986+0200	debug	fileconsumer/file.go:167	Consuming files{paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]}	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:56.185+0200	debug	fileconsumer/file.go:135	matched files%!(EXTRA zapcore.Field={paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]})	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:56.185+0200	debug	fileconsumer/file.go:167	Consuming files{paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]}	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:56.386+0200	debug	fileconsumer/file.go:135	matched files%!(EXTRA zapcore.Field={paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]})	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:56.386+0200	debug	fileconsumer/file.go:167	Consuming files{paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]}	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:56.586+0200	debug	fileconsumer/file.go:135	matched files%!(EXTRA zapcore.Field={paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]})	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-01-26T12:59:56.586+0200	debug	fileconsumer/file.go:167	Consuming files{paths 1 0  [/var/log/busybox/bar.log /var/log/busybox/bzr.log /var/log/busybox/foo.1.log /var/log/busybox/foo.log /var/log/busybox/foo1.log /var/log/busybox/log3.log /var/log/busybox/simple.log /var/log/busybox/some.log]}	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
```

Sample metrics:

```console
curl http://localhost:8888/metrics
# HELP otelcol_exporter_queue_capacity Fixed capacity of the retry queue (in batches)
# TYPE otelcol_exporter_queue_capacity gauge
otelcol_exporter_queue_capacity{exporter="otlp/elastic",service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 1000
# HELP otelcol_exporter_queue_size Current size of the retry queue (in batches)
# TYPE otelcol_exporter_queue_size gauge
otelcol_exporter_queue_size{exporter="otlp/elastic",service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 0
# HELP otelcol_exporter_sent_log_records Number of log record successfully sent to destination.
# TYPE otelcol_exporter_sent_log_records counter
otelcol_exporter_sent_log_records{exporter="otlp/elastic",service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 11
# HELP otelcol_process_cpu_seconds Total CPU user and system time in seconds
# TYPE otelcol_process_cpu_seconds counter
otelcol_process_cpu_seconds{service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 0.76
# HELP otelcol_process_memory_rss Total physical memory (resident set size)
# TYPE otelcol_process_memory_rss gauge
otelcol_process_memory_rss{service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 1.84000512e+08
# HELP otelcol_process_runtime_heap_alloc_bytes Bytes of allocated heap objects (see 'go doc runtime.MemStats.HeapAlloc')
# TYPE otelcol_process_runtime_heap_alloc_bytes gauge
otelcol_process_runtime_heap_alloc_bytes{service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 4.588912e+07
# HELP otelcol_process_runtime_total_alloc_bytes Cumulative bytes allocated for heap objects (see 'go doc runtime.MemStats.TotalAlloc')
# TYPE otelcol_process_runtime_total_alloc_bytes counter
otelcol_process_runtime_total_alloc_bytes{service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 1.40668584e+08
# HELP otelcol_process_runtime_total_sys_memory_bytes Total bytes of memory obtained from the OS (see 'go doc runtime.MemStats.Sys')
# TYPE otelcol_process_runtime_total_sys_memory_bytes gauge
otelcol_process_runtime_total_sys_memory_bytes{service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 9.3970712e+07
# HELP otelcol_process_uptime Uptime of the process
# TYPE otelcol_process_uptime counter
otelcol_process_uptime{service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev"} 111.114353151
# HELP otelcol_receiver_accepted_log_records Number of log records successfully pushed into the pipeline.
# TYPE otelcol_receiver_accepted_log_records counter
otelcol_receiver_accepted_log_records{receiver="filelog",service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev",transport=""} 11
# HELP otelcol_receiver_refused_log_records Number of log records that could not be pushed into the pipeline.
# TYPE otelcol_receiver_refused_log_records counter
otelcol_receiver_refused_log_records{receiver="filelog",service_instance_id="f9da8915-5bbd-444f-84cf-60a42107a83e",service_name="otelcontribcol",service_version="0.91.0-dev",transport=""} 0
```

**Comment by djaglowski** [comment]
> Sufficient logging when files are discovered, moved, truncated, etc

We have some of this already and the rest should be reasonable to add. However, there are some caveats.
- Discovered: [already being logged](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/31d5f56ede4244da6efe3c895eacce005f06bc16/pkg/stanza/fileconsumer/file.go#L261), but we maintain a limited memory of the files we've previously seen.
- Moved: We don't currently log this but we should be able to detect it by crossreferencing file names, which I think we have in the necessary places. It would add some computation so perhaps this should be configurable.
- Truncated: [We currently "detect" this but do not log](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/31d5f56ede4244da6efe3c895eacce005f06bc16/pkg/stanza/fileconsumer/file_other.go#L31-L34). I think the logic is not perfect and it is possible to miss truncations depending on timing. It'd be helpful to have other perspectives on how this could work performantly.

> Observable metrics about its current state: how many files are opened, being harvested, stale, etc

What's your definition of stale?

Up-down counters for open & reading should be straightforward enough to maintain.

**Comment by ChrsMark** [comment]
> What's your definition of stale?

`stale` when we have reached `EOF` and the receiver would wait until it looks again for new lines. I'm not sure if this behavior exists though in the receiver. If not, then we can just ignore this metric.

**Comment by djaglowski** [comment]
> `stale` when we have reached `EOF` and the receiver would wait until it looks again for new lines. I'm not sure if this behavior exists though in the receiver. If not, then we can just ignore this metric.

This is pretty much the same as open less reading, but maybe there's some more nuance. We can consider when implementing.

In any case, I'm marking this issue as required for ga because I think these are very basic observability signals which we should have in place.

**Comment by ChrsMark** [comment]
Heads up on this. https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31544 has been merged. That one provides the metrics about the `open` and `reading` files. The `stale` files are essentially the `open` minus the `reading` ones, so we can avoid explicitly reporting it. We can revisit this if we have any requests for that.

What is remaining from this issue now is the logging enhancements. I will be working on this soon.

**Comment by ChrsMark** [comment]
> Moved: We don't currently log this but we should be able to detect it by crossreferencing file names, which I think we have in the necessary places. It would add some computation so perhaps this should be configurable.
Truncated: [We currently "detect" this but do not log](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/31d5f56ede4244da6efe3c895eacce005f06bc16/pkg/stanza/fileconsumer/file_other.go#L31-L34). I think the logic is not perfect and it is possible to miss truncations depending on timing. It'd be helpful to have other perspectives on how this could work performantly.

Hey @djaglowski, I tried my luck on the `fileconsumer` package to explore where we handle the `truncate` and `move` cases.
Based on your [comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31256#issuecomment-1944225866) as well, we handle the truncates at https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.101.0/pkg/stanza/fileconsumer/file_other.go#L41-L43.
However, as the comment in the code also implies, if an `oldreader` doesn't point to a truncated file then it points to a moved file.
Hence, we know that a file is moved after we pass the `oldReader.Validate()` check and we can log it.

Did you have something additional in mind when mentioning the following?

> Moved: We don't currently log this but we should be able to detect it by crossreferencing file names, which I think we have in the necessary places. It would add some computation so perhaps this should be configurable.

**Comment by djaglowski** [comment]
When we detect [this condition](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c0b51136575aa7ba89326d18edb4549e7e1bbdb9/pkg/stanza/fileconsumer/file_other.go#L29) and continue, we're not concerned with whether or not a file has been rotated. It may have, or may not have. The goal when originally written here was to find files which were rotated _out of the matching pattern_. Files which were rotated _within the matching pattern_ would not be logged if we only log based on the `Validate` function later.

**Comment by ChrsMark** [comment]
Thank's for clarifying this. I see the distinction between the "out-of-pattern" rotated files and "within-pattern" ones.

I did some extra digging into the implementation and I found that we can spot the "within-pattern" rotated files at [this if-block](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.101.0/pkg/stanza/fileconsumer/file.go#L225-L227) if we compare the fileName of the `oldreader` with that of the current `file`.

However, at this point we only know that a file has been rotated from one filePath to a new one within the matching pattern. I guess we can use the same `Validate` approach to distinguish if it's truncated or not.

I'm testing these scenarios manually with the steps I mention at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33237.
